### ISSUE TITLE
Cleanup Get[Instruction name]Offset() functions with the same copy-pasted body.

### ISSUE THREAD

**Issue by kilvadyb** [v8mips/v8mips#264]
There are a lot of Get*Offset() functions (GetLwOffset(), GetLuiOffset(), GetOriOffset(), ...) in MacroAssembler with the same body. Also we have Assembler::GetImmediate16(instr) with the same body. I would like to decrease the number of functions or at least create _protected_
Assembler::GetImmediate16(instr)
Assembler::GetImmediate19(instr)
Assembler::GetImmediate21(instr)
or even better:
Assembler::GetImmediate(instr, bitnum)
And use only that function everywhere. I had a conversation (copied below from slack) with Miran about this topic. Please read it and give some feedback, ideas if you have.

mkaric [1:52] 
Do you think it is OK to use GetImmediate16 function instead, essentially it does the same thing and there are DCHECKs before the call to the function so there is no need for DCHECK inside it.

kilvadyb [2:08 PM] 
I would prefer that personally, I don't like the 20+ alias functions with the same body but there are already some (GetLwOffset(), ...) so it seems to be a common practice and the whole team should discuss it, I think. Right now just leave that part of the code of the CL as it is  IMO. The team should agree if
- We like to browse endless list of function names.
- We don’t follow the old practice and all the new CLs from now will use the same GetImmediate<bitnum>(instr)
- We do a full and deep code cleanup to use only GetImmediate<bitnum>(instr) functions. (edited)

[2:09] 
If we use only GetImmediate<bitnum>(instr) functions then the developer must know which ‘bitnum' is good for which instruction. (edited)
Or LwOffset(), LuiOffset(), … can be an inline alias to the proper GetImmediate<bitnum>() to avoid the lot of copy-paste code duplication. Like I did for branch_offset*() functions. (edited)
We must count that many of Google guys are using [macro-]assembler-mips sources before a heavy cleanup. But it is quite internal stuff...

mkaric [2:40 PM] 
I'd rather change the code to use existing GetImmediate16 function instead of creating new functions. It would be nice to get more comments on the topic from others.

**Comment by ilijapavlovic** [comment]
I agree with this idea. In constants-mips*.h there are also similar functions:

```
inline int32_t ImmValue(int bits) const;
inline int32_t Imm16Value() const;
inline int32_t Imm18Value() const;
inline int32_t Imm19Value() const;
inline int32_t Imm21Value() const;
inline int32_t Imm26Value() const;
```

So, your list can be supplemented with GetImmediate18(inst) and GetImmediate26(instr).

**Comment by kilvadyb** [comment]
Thanks, nice catch. 
-  ImmXXValue() functions can be used directly and we can remove all the GetImmediateXX() and Get*Offset() functions.
- We can make an inline alias `Assembler::GetImmediate(instr, bitnum) { return instr->ImmValue(bitnum); }`
- We can make Get*Offset()s as aliases: Get*Offset(instr) { return instr->ImmValue(proper_bitnum); }

In the last case the developer don't have to know the bit depth of an offset. But I think when we are implementing some low-level stuff and for example offset of an `lw` is neede then we must know already what bit depth to expect as a result.

Please vote :)

I think we should clean this up as inline or not inline versions of

```
int16_t *(Instr instr) {
  return ((instr & kImm16Mask));
}
```

comes up so many times. For example if we decide to treat all the offset values (from 16 to 26 bits variants) uniformly as int32_t then 20+ versions of this function should be changed.

**Comment by ilijapavlovic** [comment]
The proposal nr. 1: - direct using ImmXXValue() functions and removing GetImmediateXX() and Get*Offset().
